### ISSUE TITLE
Classes in test source folder being packaged into WAR

### ISSUE THREAD

**Issue by chanseokoh** [GoogleCloudPlatform/google-cloud-eclipse#1997]
1. Create a new _native_ HelloWorld project.
2. Move `HelloWorld.java` to `src/test/java`. Optionally delete `src/main/java` to be sure.
3. Run locally or deploy.

Note that the `HelloWorld` servlet is accessible.

**Comment by chanseokoh** [comment]
Not verified, but probably this won't happen to Maven standard projects; the output folder of `src/test/java` in the Maven projects are set to `target/test-classes` and the deployment assembly setting does not pick up that directory:

![selection_020](https://cloud.githubusercontent.com/assets/10523105/26699761/e4cfa89e-46e8-11e7-82d6-f90152b99e00.png)

![selection_021](https://cloud.githubusercontent.com/assets/10523105/26699769/ec59fbfa-46e8-11e7-845c-5f6b9217b3b9.png)

**Comment by chanseokoh** [comment]
Two things should be done:
1. Remove the `/src/test/java` entry from the _Web Deployment Assembly_ sources.
1. Set the output folder of `src/test/java` to `build/test-classes` (for non-Maven projects) or `target/test-classes` (for Maven projects).

**Comment by chanseokoh** [comment]
Re-opening as sometimes I still see `/src/test/java` being assembled:

![selection_030](https://user-images.githubusercontent.com/10523105/26944615-361a4de4-4c57-11e7-9c49-024dbaa54759.png)

If this happens, it usually happens when I create a project for the first time in a hosted debug run, where the wizards take some time while loading things for the first time.

This also seems to result in occasional test failure. From https://github.com/GoogleCloudPlatform/google-cloud-eclipse/pull/2033#issuecomment-307185425:

```
Tests run: 9, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 84.646 sec <<< FAILURE! - in com.google.cloud.tools.eclipse.appengine.newproject.standard.CreateAppEngineStandardWtpProjectTest

testMostImportantFile(com.google.cloud.tools.eclipse.appengine.newproject.standard.CreateAppEngineStandardWtpProjectTest)  Time elapsed: 31.902 sec

testAppEngineLibrariesAdded(com.google.cloud.tools.eclipse.appengine.newproject.standard.CreateAppEngineStandardWtpProjectTest)  Time elapsed: 17.613 sec

testNoTestClassesInDeploymentAssembly(com.google.cloud.tools.eclipse.appengine.newproject.standard.CreateAppEngineStandardWtpProjectTest)  Time elapsed: 7.521 sec  <<< FAILURE!

java.lang.AssertionError: null
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertFalse(Assert.java:64)
	at org.junit.Assert.assertFalse(Assert.java:74)
	at com.google.cloud.tools.eclipse.appengine.newproject.standard.CreateAppEngineStandardWtpProjectTest.assertNoTestClassesInDeploymentAssembly(CreateAppEngineStandardWtpProjectTest.java:212)
	at com.google.cloud.tools.eclipse.appengine.newproject.standard.CreateAppEngineStandardWtpProjectTest.testNoTestClassesInDeploymentAssembly(CreateAppEngineStandardWtpProjectTest.java:201)
```

**Comment by chanseokoh** [comment]
I've spent some time on https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/1997#issuecomment-307189699, but so far no luck. I am going to disable the flaky test temporarily until this is fixed.

**Comment by briandealwis** [comment]
Maybe the ~~solution~~ workaround for #1997 is to *not* add `src/test/java` as a source path in `FacetUtil#addJavaFacetToBatch()` and instead add it in [`CreateAppEngineWtpProject`](https://github.com/GoogleCloudPlatform/google-cloud-eclipse/pull/2005/files#diff-4b392ccec6b10d92b2818dab889fbfdbR123)?  Or better yet (perhaps) add a facet-change listener in `FacetUtil#addJavaFacetToBatch()` to do so?

**Comment by chanseokoh** [comment]
I can't reproduce the occasional failure locally today. Maybe it happens when a system is under heavy load. Since I'm having difficulty diagnosing without being able to reproduce it, I'll go with the workaround @briandealwis said, which will be a definite end to this subtle failure.

**Comment by chanseokoh** [comment]
@briandealwis unfortunately, the workaround is not going to work; it doesn't matter if we add `src/test/java` later or not. Adding any src folder seems to (or is supposed to) trigger the listener `org.eclipse.jst.j2ee.refactor.listeners.J2EEElementChangedListener`, which adds the src folder to deployment assembly. This may explain the occasional failure we've seen so far; even if we explicitly removed `src/test/java` from deploy assembly by calling `WebProjectUtil.removeWebDeploymentAssemblyEntry()`, changing the output folder of `src/test/java` may trigger the listener to add `src/test/java` back to deployment assembly.

Now, what is extremely weird is that, although I think changing Java classpath entries is supposed to trigger the listener all the time, it rarely does so in practice in our case. (And when it does, it causes the occasional failures we've seen so far.) It doesn't make sense at all that it is triggered some times and not at all in other times.

**Comment by chanseokoh** [comment]
I found a way to reproduce this.

1. Set a breakpoint [here](https://github.com/GoogleCloudPlatform/google-cloud-eclipse/blob/v1.2.0B/plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/CreateAppEngineWtpProject.java#L148).
1. Start a hosted debug run. Create a new project (std or flex).
1. You'll hit the breakpoint. Press F8 to continue, and you're done.

`J2EEElementChangedListener.processJavaProject()` will be executed, and `src/test/java` will be added back by `createLink()` below:
```java
					for(int i=0;i<pathsToAdd.size(); i++){
						Object[] toAdd = (Object[]) pathsToAdd.get(i);
						final IVirtualFolder destFolder = (IVirtualFolder) toAdd[1];
						final IPath pathToAdd = (IPath) toAdd[0];
						destFolder.createLink(pathToAdd, 0, monitor);
					}
```

**Comment by chanseokoh** [comment]
Although now I can reproduce this, I can't think of any good reason why sometimes the listener is not executed (or is executed). For now, I'll unassign myself from this issue and work on other pressing issues.

**Comment by briandealwis** [comment]
m2e-wtp must have this problem too.  Looking at that code:
  - [WebProjectConfiguratorDelegate](http://git.eclipse.org/c/m2e-wtp/org.eclipse.m2e.wtp.git/tree/org.eclipse.m2e.wtp/src/org/eclipse/m2e/wtp/WebProjectConfiguratorDelegate.java?h=master#n149) calls a helper method [`WTPProjectsUtil.remoteTestFolderLinks()`](http://git.eclipse.org/c/m2e-wtp/org.eclipse.m2e.wtp.git/tree/org.eclipse.m2e.wtp/src/org/eclipse/m2e/wtp/WTPProjectsUtil.java?h=master#n513):
>     // MNGECLIPSE-632 remove test sources/resources from WEB-INF/classes
>     removeTestFolderLinks(project, mavenProject, monitor, "/WEB-INF/classes"); //$NON-NLS-1$
> 
>     addContainerAttribute(project, DEPENDENCY_ATTRIBUTE, monitor);

The `DEPENDENCY_ATTRIBUTE` references a [WTP wiki page](https://wiki.eclipse.org/ClasspathEntriesPublishExportSupport) that discusses publishing/exporting classpath dependencies.  It says associated a `org.eclipse.jst.component.nondependency` classpath attribute to a classpath entry will prevent it from being exported.

But that said, the `addContainerAttribute()` call only changes classpath containers, not source entries.

I haven't tried running or digging deeper into this.

**Comment by chanseokoh** [comment]
Removing the test folder link itself works well. The problem here is that our effort to remove the link is later overwritten by the test source classpath change. I wish we could know when it will be overwritten and remove the link after that change.

> But that said, the addContainerAttribute() call only changes classpath containers, not source entries.

I also tried setting this attribute to the test source classpath, but it doesn't seem to work.

**Comment by chanseokoh** [comment]
Another way to reproduce is to put `Thread.sleep()` at the following location. (However, setting a breakpoint is a more reliable way.)

```java
  private static void fixTestSourceDirectorySettings(IProject newProject, IProgressMonitor monitor)
      throws CoreException {
    // 1. Fix the output folder of "src/test/java".
    IPath testSourcePath = newProject.getFolder("src/test/java").getFullPath();

    IJavaProject javaProject = JavaCore.create(newProject);
    IClasspathEntry[] entries = javaProject.getRawClasspath();
    try {
      Thread.sleep(5000);
    } catch (InterruptedException e) {}
    for (int i = 0; i < entries.length; i++) {
      IClasspathEntry entry = entries[i];
```

**Comment by chanseokoh** [comment]
I think I found one workaround the issue that sometimes the `J2EEElementChangedListener` is (not) notified of the classpath change: wrap `CreateAppEngineWtpProject` in `IWorkspace.run()` to effectively suppress some interim and unnecessary project change events that may be fired during project creation:

```java
    try {
      ResourcesPlugin.getWorkspace().run(new IWorkspaceRunnable() {
        @Override
        public void run(IProgressMonitor monitor) throws CoreException {
          boolean fork = false;
          boolean cancelable = true;
          try {
            getContainer().run(fork, cancelable, runnable);
          } catch (InvocationTargetException | InterruptedException e) {}
        }
      }, new NullProgressMonitor());

      // open most important file created by wizard in editor
      IFile file = runnable.getMostImportant();
      WorkbenchUtil.openInEditor(workbench, file);
      return true;
    } catch (CoreException ex) {
      return false;
    }
```

This way, our interim classpath change (i.e., changing the output directory of "src/test/java") will not fire an event saying that some classpath in a project has changed. (In the end, the workspace will just see that a project has been created with given classpaths.)

**Comment by chanseokoh** [comment]
So, here's my theory:

`CreateAppEngineWtpProject` implements `WorkspaceModifyOperation`. According to its Javadoc, it says it generally defers resource change events, but it may still broadcast periodic resource change notifications, e.g., if the operation takes a long time:

```javadoc
 * modification should be performed using this operation. The primary
 * consequence of using this operation is that events which typically occur as a
 * result of workspace changes (such as the firing of resource deltas,
 * performance of autobuilds, etc.) are generally deferred until the outermost operation
 * has successfully completed.  The platform may still decide to broadcast
 * periodic resource change notifications during the scope of the operation
 * if the operation runs for a long time or another thread modifies the workspace
 * concurrently.
```

Given https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/1997#issuecomment-308175999, https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/1997#issuecomment-309881606, and that

> If this happens, it usually happens when I create a project for the first time in a hosted debug run, where the wizards take some time while loading things for the first time.

I think our `CreateAppEngineWtpProject` operation may fire an classpath change event if the whole operation takes a long time. And if that happens, `J2EEElementChangedListener` is notified that the classpath `src/test/java` has been changed (described as REMOVED and ADDED) and adds the path to the deploy assembly. (By its design, the job of `J2EEElementChangedListener` is to add a source path to the deploy assembly whenever it is notified that the path has been added to a project.)

**Comment by chanseokoh** [comment]
FYI, `WorkspaceModifyOperation.run()` runs a given operation via `IWorkspace.run()` too, but when calling it, it does not set the `IWorkspace.AVOID_UPDATE` which suppresses update events.

**Comment by chanseokoh** [comment]
BTW, this happened yesterday on Kokoro (https://github.com/GoogleCloudPlatform/google-cloud-eclipse/pull/2073#issuecomment-309528661), which runs Oxygen RC2.

**Comment by chanseokoh** [comment]
However, looks like there is no guarantee in https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/1997#issuecomment-309887188 that it will prevent all notifications.

https://eclipse.org/articles/Article-Resource-deltas/resource-deltas.html

> Since Eclipse 3.0, it is no longer guaranteed that an IWorkspaceRunnable will prevent notifications for the entire duration of an operation. The workspace can now decide to perform notifications during an operation to ensure UI responsiveness. This is particularly important when several workspace modifying operations are running simultaneously. The use of IWorkspaceRunnable is still strongly encouraged, functioning as a strong hint to the workspace that a set of changes is occurring that can be batched.

**Comment by chanseokoh** [comment]
I thought about a few workarounds, and I think the following is the best among them (although not ideal):

1. Add the Java facet (with or without "src/test/java" as a source folder).
1. Fix the output directory of "src/test/java" (or add it with the proper output directory if we didn't add it above.)
1. Add the WTP facet. This will add "src/test/java" to the deploy assembly.
1. Remove "src/test/java" from the deploy assembly.

This way, because we make changes to the "src/test/java" classpath before adding the WTP facet, there is absolutely no possibility that `J2EEElementChangedListener` will respond to the change.

The drawback is that we should break the `FacetUtil`'s batch mechanism when adding all the facets: Java, WTP, and App Engine. (The reason for the batch was to fix #1155, and fortunately, we only need to batch WTP and App Engine together for #1155.)

**Comment by chanseokoh** [comment]
@briandealwis got a better idea. We believe that, if changing a classpath entry fires an event, it should synchronously call any registered listeners, including `J2EEElementChangedListener`. (The `J2EE*Listener` will then schedule a job.) So, we just need to wait for the job!

There is a problem of deadlock when waiting (because our project creation operation holds a workspace lock and the job schedule by `J2EE*Listener` also try to modify a resource in the workspace), so we need a workaround like waiting for the job in another job.
